internal/prefixcollector/config_map_prefix_source.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prefixcollector
+
+import (
+	"cmd-exclude-prefixes-k8s/internal/utils"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const configMapPrefixesKey = "excluded_prefixes.yaml"
+
+// ConfigMapPrefixSource is Kubernetes ConfigMap excluded prefix source
+type ConfigMapPrefixSource struct {
+	configMapName      string
+	configMapNameSpace string
+	configMapInterface v1.ConfigMapInterface
+	prefixes           utils.SynchronizedPrefixesContainer
+	ctx                context.Context
+}
+
+// NewConfigMapPrefixSource creates ConfigMapPrefixSource
+func NewConfigMapPrefixSource(ctx context.Context, notify *sync.Cond, name, namespace string) *ConfigMapPrefixSource {
+	clientSet := FromContext(ctx)
+	configMapInterface := clientSet.CoreV1().ConfigMaps(namespace)
+	cmps := ConfigMapPrefixSource{
+		configMapName:      name,
+		configMapNameSpace: namespace,
+		configMapInterface: configMapInterface,
+		ctx:                ctx,
+	}
+
+	go cmps.watchConfigMap(notify)
+	return &cmps
+}
+
+// Prefixes returns prefixes from source
+func (cmps *ConfigMapPrefixSource) Prefixes() []string {
+	return cmps.prefixes.GetList()
+}
+
+func (cmps *ConfigMapPrefixSource) watchConfigMap(notify *sync.Cond) {
+	for {
+		cm, err := cmps.configMapInterface.Get(cmps.ctx, cmps.configMapName, metav1.GetOptions{})
+		if err != nil {
+			logrus.Errorf("Failed to get ConfigMap '%sources/%sources': %v", cmps.configMapNameSpace, cmps.configMapName, err)
+			return
+		}
+
+		bytes := []byte(cm.Data[configMapPrefixesKey])
+		prefixes, err := utils.YamlToPrefixes(bytes)
+		if err != nil {
+			logrus.Errorf("Can not unmarshal prefixes, err: %v", err.Error())
+			return
+		}
+		cmps.prefixes.SetList(prefixes.PrefixesList)
+		notify.Broadcast()
+		<-time.After(time.Second * 10)
```

Comment:
Can we use configMapInterface.Watch here instead of polling?

Reply:
Fixed.